File: backend/app/derive.py
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def parse_iso(ts: str) -> datetime:
    # Accept ISO 8601 with timezone; fall back to UTC naive as UTC.
    dt = datetime.fromisoformat(ts.replace("Z", "+00:00"))
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt


def read_events(events_path: Path) -> list[dict[str, Any]]:
    if not events_path.exists():
        return []
    events: list[dict[str, Any]] = []
    with events_path.open("r", encoding="utf-8") as handle:
        for line in handle:
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                # v0 stance: ignore malformed lines rather than fail the whole app.
                continue
            if isinstance(obj, dict):
                events.append(obj)
    return events


@dataclass(frozen=True)
class DerivedTaskSignals:
    last_done_ts: datetime | None
    last_skip_ts: datetime | None
    skip_count_recent: int
    scan_boost: float


def derive_task_signals(
    *,
    now: datetime,
    task_id: str,
    room_context: str,
    linked_check_ids: set[str],
    events: list[dict[str, Any]],
    scan_boost_window_hours: float = 48.0,
    scan_boost_decay_hours: float = 48.0,
    skip_window_days: int = 7,
) -> DerivedTaskSignals:
    last_done_ts: datetime | None = None
    last_skip_ts: datetime | None = None
    skip_count_recent = 0
    scan_boost = 0.0

    skip_cutoff = now.timestamp() - (skip_window_days * 86400)
    scan_cutoff = now.timestamp() - (scan_boost_window_hours * 3600)

    for event in events:
        etype = event.get("type")
        ets_raw = event.get("ts")
        if not isinstance(ets_raw, str):
            continue
        try:
            ets = parse_iso(ets_raw)
        except ValueError:
            continue

        if etype == "task_done" and event.get("task_id") == task_id:
            if last_done_ts is None or ets > last_done_ts:
                last_done_ts = ets

        if etype == "task_skip" and event.get("task_id") == task_id:
            if last_skip_ts is None or ets > last_skip_ts:
                last_skip_ts = ets
            if ets.timestamp() >= skip_cutoff:
                skip_count_recent += 1

        if etype == "scan_answer":
            if event.get("room") != room_context:
                continue
            if event.get("answer") != "no":
                continue
            check_id = event.get("check_id")
            if not isinstance(check_id, str) or check_id not in linked_check_ids:
                continue
            if ets.timestamp() < scan_cutoff:
                continue
            age_hours = max(0.0, (now - ets).total_seconds() / 3600.0)
            decay = max(0.0, 1.0 - (age_hours / scan_boost_decay_hours))
            scan_boost += decay

    return DerivedTaskSignals(
        last_done_ts=last_done_ts,
        last_skip_ts=last_skip_ts,
        skip_count_recent=skip_count_recent,
        scan_boost=scan_boost,
    )


def overdue_factor(*, now: datetime, last_done_ts: datetime | None, frequency_days: int | None) -> float:
    if not frequency_days:
        return 0.0
    if last_done_ts is None:
        return 1.0
    age_days = (now - last_done_ts).total_seconds() / 86400.0
    return max(0.0, age_days / float(frequency_days))

File: backend-rs/src/logic/derive.rs
use chrono::{DateTime, Duration, Utc};
use serde_json::Value;
use std::collections::HashSet;
use std::path::Path;

pub fn parse_event_ts(ts: &str) -> Option<DateTime<Utc>> {
    // Accept RFC 3339 with or without sub-second precision, with Z or +00:00 offset.
    if let Ok(dt) = DateTime::parse_from_rfc3339(ts) {
        return Some(dt.with_timezone(&Utc));
    }
    // Fallback: allow naive-ish ISO strings by assuming UTC.
    if let Ok(dt) = chrono::NaiveDateTime::parse_from_str(ts, "%Y-%m-%dT%H:%M:%S%.f") {
        return Some(dt.and_utc());
    }
    None
}

pub fn read_events(path: &Path) -> Vec<Value> {
    if !path.exists() {
        return Vec::new();
    }
    let content = match std::fs::read_to_string(path) {
        Ok(s) => s,
        Err(_) => return Vec::new(),
    };
    content
        .lines()
        .filter(|line| !line.trim().is_empty())
        .filter_map(|line| serde_json::from_str(line).ok())
        .collect()
}

#[derive(Debug, Clone)]
pub struct DerivedTaskSignals {
    pub last_done_ts: Option<DateTime<Utc>>,
    pub last_skip_ts: Option<DateTime<Utc>>,
    pub skip_count_recent: i32,
    pub scan_boost: f64,
}

#[allow(clippy::too_many_arguments)]
pub fn derive_task_signals(
    now: DateTime<Utc>,
    task_id: &str,
    room_context: &str,
    linked_check_ids: &HashSet<String>,
    events: &[Value],
    scan_boost_window_hours: f64,
    scan_boost_decay_hours: f64,
    skip_window_days: i32,
) -> DerivedTaskSignals {
    let mut last_done_ts: Option<DateTime<Utc>> = None;
    let mut last_skip_ts: Option<DateTime<Utc>> = None;
    let mut skip_count_recent = 0;
    let mut scan_boost = 0.0;

    let skip_cutoff = now - Duration::days(skip_window_days as i64);
    let scan_cutoff = now - Duration::minutes((scan_boost_window_hours * 60.0) as i64);

    for event in events {
        let etype = event.get("type").and_then(|v| v.as_str());
        let ets_raw = event.get("ts").and_then(|v| v.as_str());

        let ets = match ets_raw {
            Some(ts) => match ts.parse::<DateTime<Utc>>() {
                Ok(dt) => dt,
                Err(_) => continue,
            },
            None => continue,
        };

        match etype {
            Some("task_done") if event.get("task_id").and_then(|v| v.as_str()) == Some(task_id)
                && last_done_ts.is_none_or(|prev| ets > prev) =>
            {
                last_done_ts = Some(ets);
            }
            Some("task_skip") if event.get("task_id").and_then(|v| v.as_str()) == Some(task_id) => {
                if last_skip_ts.is_none_or(|prev| ets > prev) {
                    last_skip_ts = Some(ets);
                }
                if ets >= skip_cutoff {
                    skip_count_recent += 1;
                }
            }
            Some("scan_answer") => {
                if event.get("room").and_then(|v| v.as_str()) != Some(room_context) {
                    continue;
                }
                if event.get("answer").and_then(|v| v.as_str()) != Some("no") {
                    continue;
                }
                let check_id = match event.get("check_id").and_then(|v| v.as_str()) {
                    Some(id) => id,
                    None => continue,
                };
                if !linked_check_ids.contains(check_id) {
                    continue;
                }
                if ets < scan_cutoff {
                    continue;
                }

                let age_hours = (now - ets).num_seconds() as f64 / 3600.0;
                let decay = (1.0 - (age_hours / scan_boost_decay_hours)).max(0.0);
                scan_boost += decay;
            }
            _ => {}
        }
    }

    DerivedTaskSignals {
        last_done_ts,
        last_skip_ts,
        skip_count_recent,
        scan_boost,
    }
}

pub fn overdue_factor(
    now: DateTime<Utc>,
    last_done_ts: Option<DateTime<Utc>>,
    frequency_days: Option<i32>,
) -> f64 {
    match frequency_days {
        None => 0.0,
        Some(freq) => {
            let last_ts = match last_done_ts {
                None => return 1.0,
                Some(ts) => ts,
            };
            let age_days = (now - last_ts).num_seconds() as f64 / 86400.0;
            (age_days / freq as f64).max(0.0)
        }
    }
}
